### PR TITLE
Allow Records with Matching Name + Type

### LINKED ISSUE #14: The input `records` should default to an empty list instead of `null`
Found a bug? Maybe our [Slack Community](https://slack.cloudposse.com) can help. 

[Slack Community](https://slack.cloudposse.com)

## Describe the Bug
The input `records` should default to an empty list instead of `null`. Currently when planning without passing in `records`, it will fail when trying to run `length(var.records)` within the module

## Expected Behavior
Allow planning without overriding the `records` input

### LINKED ISSUE #15: Improve `records` interface to allow for multiple records with unique values
Have a question? Please checkout our [Slack Community](https://slack.cloudposse.com) or visit our [Slack Archive](https://archive.sweetops.com/). 

[Slack Community](https://slack.cloudposse.com)

## Describe the Feature

Need to create 4 records with the same value with different targets

i.e.

```
test.example.com NS aws-resolver1.blabla
test.example.com NS aws-resolver2.blabla
test.example.com NS aws-resolver21.blabla
test.example.com NS aws-resolver88.blabla
```

This results in resource conflicts because the `value` is not added to the key

```
test.example.comNS
test.example.comNS
test.example.comNS
test.example.comNS
```

## Expected Behavior

Unique keys per record

## Proposal

```hcl
  records = [
    {
      key   = "bastion_A"
      name  = "bastion"
      value = "192.168.1.11"
      type  = "A"
      ttl   = 3600
    },
    {
      key   = "api_A"
      name  = "api"
      value = "192.168.2.22"
      type  = "A"
      ttl   = 3600
    },
    {
      key   = "test.example.com_NS_1"
      name  = "test"
      value = "aws-resolver1.blabla"
      type  = "NS"
      ttl   = 3600
    },
    {
      key   = "test.example.com_NS_2"
      name  = "test"
      value = "aws-resolver2.blabla"
      type  = "NS"
      ttl   = 3600
    },
  ]
```

This would

- Allow `key` to be optional but if it was set, it would be mandatory for all items in the list
- Backwards compatibility

## references

- https://sweetops.slack.com/archives/CB6GHNLG0/p1660906495013409

### PR DESCRIPTION
**Pull request by milldr**
## what
- Allowing for records to have the same Name and Type
- Added index to key for records map

## why
- We want to be able to create records that have the same name and type, so that we can create NS (different values)
- It is up to the user to ensure that duplicate records are not attempted to be created. Otherwise we would need to add name+type+_value_ as key for the record map

## references
- Closes #15 
- Closes #14

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -6,11 +6,14 @@ locals {
...
+
+  flatten_records = flatten(var.records)
   records = local.records_enabled ? {
-    for record in flatten(var.records) :
-    format("%s%s",
+    for record in local.flatten_records :
+    format("%s%s%s",
       record.name,
       record.type,
+      index(local.flatten_records, record)
```
Comment (nitrocode) [inline main.tf:16]: It's probably obvious but could you explain what the value of the index function would return as an example of a record input to this module?

**Review by nitrocode — CHANGES_REQUESTED** [review]
Thank you for putting in this PR.

@Nuru and I spoke of this change not too long ago and we decided that a key input would be the best path forward.

A few days ago we received this PR which does have the key input. It's worth reviewing that one instead of this one.

https://github.com/cloudposse/terraform-cloudflare-zone/pull/17/files

**Comment by milldr** [comment]
Closing in favor of #17

**Pull request closed without merging**